engines/afalg/e_afalg.c

```diff
@@ -230,7 +230,11 @@ int afalg_fin_cipher_aio(afalg_aio *aio, int sfd, unsigned char *buf,
     memset(cb, '\0', sizeof(*cb));
     cb->aio_fildes = sfd;
     cb->aio_lio_opcode = IOCB_CMD_PREAD;
-    cb->aio_buf = (uint64_t)buf;
+    /*
+     * The pointer has to be converted to unsigned value first to avoid
+     * sign extension on cast to 64 bit value
+     */
+    cb->aio_buf = (uint64_t)(unsigned long)buf;
```

Comment:
Thinking about this more, maybe something like the following would be even better. It avoids getting tangled up in data models (like `ILP32`) and platforms (like X32 and Aarch32).

```
#if sizeof(u64) != sizeof(cb->aio_buf)
cb->aio_buf = (uint64_t)(unsigned long)buf;
#else
cb->aio_buf = (uint64_t)buf;
#endif
```